### PR TITLE
[FEA] Introduce low shuffle merge.

### LINKED ISSUE #10905: [FEA] Introduce `GpuLowShuffleMergeCommand`.

### PR DESCRIPTION
**Pull request by liurenjie1024**
This pr implements low shuffle merge algorithm to improve merge performance, but currently it has some limitations:

1. It only support databricks 13.3 and delta lake 24x.
2. It requires that the scan must be PERFILE.
3. Currently the deletion vector generation is using cpu, we can improve this by using gpu accelated version.

Close #10905

### CODE REVIEW

File: jenkins/spark-premerge-build.sh [hunk 1]
```diff
@@ -206,7 +206,7 @@ ci_scala213() {
     cd .. # Run integration tests in the project root dir to leverage test cases and resource files
     export TEST_TAGS="not premerge_ci_1"
     export TEST_TYPE="pre-commit"
-    export TEST_PARALLEL=5
```
Comment (NvTimLiu) [inline jenkins/spark-premerge-build.sh:209]: @liurenjie1024  Can you explain why we need to change PARALLEL to 4 from 5?


Please note : this change may make our pre-merge build become slower

Comment (NvTimLiu) [reply jenkins/spark-premerge-build.sh:209]: > This is to workaround this bug: https://github.com/NVIDIA/spark-rapids/issues/8652

Thanks for your info @liurenjie1024 

I did not found connections between this PR and issue/8652, can you help explain it, thanks a lot!

BTW, if the change here `TEST_PARALLEL=4` is a workaround instead fixing, can you help add notation like `//TODOL revert the line change after issue 652 get fixed` ahead of the changing line? 

Please note : this change is critical, it makes our the pre-merge CI pipeline become slower.
It will impact all the coming pre-merge CI if this PR get merged!

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormatUtils.scala [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    val delVecCol = delVec.map { delVec =>
+      withResource(Scalar.fromBool(false)) { s =>
+        withResource(CudfColumnVector.fromScalar(s, batch.numRows())) { c =>
+          var table = new Table(c)
+          val posIter = new RoaringBitmapIterator(
+            delVec.getLongIteratorFrom(rowIdxStart),
+            rowIdxStart,
+            rowIdxStart + batch.numRows(),
+          ).grouped(Math.min(maxBatchSize, batch.numRows()))
```
Comment (jlowe) [inline delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormatUtils.scala:106]: Curious why we would iterate in smaller batches?  We're asking the GPU to do tiny amounts of work multiple times by using a small batch size here.  Not sure why this is configurable and set to a relatively small batch size.

**Review by jlowe — COMMENTED** [review]
I've only had time to skim some parts of this, adding comments I have so far.

File: pom.xml [hunk 3]
```diff
@@ -922,6 +922,7 @@
         <bloop.configDirectory>${spark.rapids.source.basedir}/.bloop</bloop.configDirectory>
         <build.info.path>${project.build.outputDirectory}/rapids4spark-version-info.properties</build.info.path>
         <nonfail.errors.quiet>false</nonfail.errors.quiet>
+        <roaringbitmap.version>1.0.6</roaringbitmap.version>
```
[the remark this answers is not in the document]
Comment (jlowe) [reply pom.xml:925]: Can you elaborate on the concerns?  We're building for very specific Spark versions and should know which version of the dependency we're getting.  I have more concerns about including it, as there's already a version provided by Spark and the one we provide can conflict with that version, breaking either us or Spark, depending upon classpath ordering.  If we do end up including this, we need to shade it and update NOTICE-binary.

Note that we're already pulling in the Spark version of RoaringBitmap in RapidsPushBasedFetchHelper and RapidsShuffleBlockFetcherIterator even before this change.

File: delta-lake/delta-spark330db/pom.xml
[the remark this answers is not in the document]
Comment (jlowe) [reply delta-lake/delta-spark330db/pom.xml:1]: If the compile failure is only on databricks, then that means we need to fix the databricks compilation environment rather than explicitly pull in a dependency that could conflict with the version found on the Databricks environment.  See shim-deps/databricks/pom.xml and jenkins/databricks/install-deps.py.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 4]
```diff
@@ -2274,6 +2274,26 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
...
+    .integerConf
+    .createWithDefault(4096)
+
+  val DELTA_LOW_SHUFFLE_MERGE_DEL_VECTOR_BROADCAST_MAX_COUNT =
+    conf("spark.rapids.sql.delta.lowShuffleMerge.deletionVector.broadcast.max.count")
+    .doc("Option to set max broadcast count of low shuffle merge deletion vector. If the " +
+        "detected deletion vector row count is larger than this value, low shuffle merge will be " +
+        "disabled.")
+    .longConf
+    .createWithDefault(600000000)
```
[the remark this answers is not in the document]
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2295]: It would be good to have a comment explaining how this was computed for future reference.  Since it was sensitive to the broadcast size, should this be derived from the configured broadcast size?

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormatUtils.scala (hunk 2 above)
Comment (jlowe) [reply delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormatUtils.scala:106]: Understand the OOM concern, but 4096 still seems really small.  That's not a lot of data for the GPU to process per iteration, and it's only 32K on the CPU.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 5]
```diff
@@ -2274,6 +2274,26 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
...
+    conf("spark.rapids.sql.delta.lowShuffleMerge.deletion.scatter.max.size")
+    .doc("Option to set max batch size when scattering deletion vector")
+    .integerConf
+    .createWithDefault(32 * 1024)
+
+  val DELTA_LOW_SHUFFLE_MERGE_DEL_VECTOR_BROADCAST_THRESHOLD =
+    conf("spark.rapids.sql.delta.lowShuffleMerge.deletionVector.broadcast.threshold")
+    .doc("""Currently we need to broadcast deletion vector to all executors to perform low
+           |shuffle merge. When we detect the deletion vector broadcast size is larger than this
+           |value, we will fallback to normal shuffle merge.""".stripMargin)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2293]: I don't think we want the stripMargin approach here, as it introduces newlines into the description which we do not want.  See the generated docs for details.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 6]
```diff
@@ -2274,6 +2274,26 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
       .integerConf
       .createWithDefault(1024)
 
+  val ENABLE_DELTA_LOW_SHUFFLE_MERGE =
+    conf("spark.rapids.sql.delta.lowShuffleMerge.enabled")
+    .doc("Option to turn on the low shuffle merge for Delta Lake.")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2279]: The documentation for this needs to elaborate on the limitations (e.g.: only available on some platforms, requires other configs to also be set, has broadcast limits, etc.).

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 7]
```diff
@@ -2274,6 +2274,26 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
...
+    .booleanConf
+    .createWithDefault(false)
+
+  val DELTA_LOW_SHUFFLE_MERGE_SCATTER_DEL_VECTOR_BATCH_SIZE =
+    conf("spark.rapids.sql.delta.lowShuffleMerge.deletion.scatter.max.size")
+    .doc("Option to set max batch size when scattering deletion vector")
+    .integerConf
+    .createWithDefault(32 * 1024)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2287]: This should be an internal config, especially if we have this done solely on the GPU which would remove the need for this.

File: jenkins/databricks/test.sh [hunk 8]
```diff
@@ -105,7 +105,7 @@ if [[ "$(pwd)" == "$SOURCE_PATH" ]]; then
 
     if [[ "$TEST_MODE" == "DEFAULT" || $TEST_MODE == "CI_PART2" || "$TEST_MODE" == "DELTA_LAKE_ONLY" ]]; then
         ## Run Delta Lake tests
-        SPARK_SUBMIT_FLAGS="$SPARK_CONF $DELTA_LAKE_CONFS" TEST_PARALLEL=1 \
+        SPARK_SUBMIT_FLAGS="$SPARK_CONF $DELTA_LAKE_CONFS" \
```
Comment (jlowe) [inline jenkins/databricks/test.sh:108]: Should revert this change, covered by #11051.

File: delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala [hunk 9]
```diff
@@ -0,0 +1,1083 @@
...
+  val INCR_ROW_COUNT_COL: String = "_incr_row_count_"
+
+  // Some Delta versions use Literal(null) which translates to a literal of NullType instead
+  // of the Literal(null, StringType) which is needed, so using a fixed version here
+  // rather than the version from Delta Lake.
+  val CDC_TYPE_NOT_CDC_LITERAL: Literal = Literal(null, StringType)
+
+  def toDeletionVector(bitmap: Roaring64Bitmap): DeletionVectorDescriptor = {
+    DeletionVectorDescriptor.inlineInLog(RoaringBitmapWrapper(bitmap).serializeToBytes(),
+      bitmap.getLongCardinality)
```
[the remark this answers is not in the document]
Comment (jlowe) [reply delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala:1064]: > Currently we can't use RoaringBitmapArray because we rely on [this api](https://javadoc.io/static/org.roaringbitmap/RoaringBitmap/0.9.25/org/roaringbitmap/longlong/Roaring64Bitmap.html#getLongIteratorFrom(long)).

I don't see why we couldn't build an interator for RoaringBitmapArray that walks the bitmaps and produces the corresponding long values.  Seems relatively straightforward, but agree not necessary for this short-term prototype, especially if we move it to the GPU.

> We are not blindly assuming it.

There is a flag for whether the bitmap is inlined or not, but the code is unconditionally calling the method to get the inlined bitmap.  If the bitmap is not inlined, the query crashes.  That's what I meant by blindly assuming.

>  The way they distinguish it is to add a 4-byte magic number as the header of bytes. We can extend this when we acutally want to support deletion vector on read.

Sounds reasonable.

**Review by jlowe — APPROVED** [review]
Thanks for all the work on this, @liurenjie1024!  Please file the followup tracking issues for this, e.g.:
- Updating Delta Lake support documentation at docs.nvidia.com
- Adding support for other Databricks and OSS Delta Lake versions
